src/main/core/work/task.rs

```diff
@@ -0,0 +1,218 @@
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+
+use crate::{
+    core::worker::Worker,
+    host::host::Host,
+    utility::{IsSend, IsSync},
+};
+
+/// Mostly for interoperability with C APIs.
+/// In Rust code that doesn't need to interact with C, it may make more sense
+/// to directly use a `FnMut(&mut Host)` trait object.
+pub struct TaskRef {
+    inner: Arc<AtomicRefCell<dyn FnMut(&mut Host) + Send + Sync>>,
+    // Runtime memory error checking to help catch errors that C code is prone
+    // to.  Can probably drop once C interop is removed, if we don't drop this
+    // type altogether.
+    #[cfg(debug_assertions)]
+    magic: u32,
+}
+
+impl TaskRef {
+    #[cfg(debug_assertions)]
+    const MAGIC: u32 = 0xe0408897;
+
+    pub fn new<T: 'static + FnMut(&mut Host) + Send + Sync>(f: T) -> Self {
+        Worker::increment_object_alloc_counter("TaskRef");
+        Self {
+            inner: Arc::new(AtomicRefCell::new(f)),
+            #[cfg(debug_assertions)]
+            magic: Self::MAGIC,
+        }
+    }
+
+    pub fn execute(&mut self, host: &mut Host) {
+        self.check_magic();
+        let mut inner = self.inner.borrow_mut();
+        inner(host)
+    }
+
+    #[cfg(debug_assertions)]
+    fn check_magic(&self) {
+        debug_assert!(self.magic == Self::MAGIC);
+    }
+    #[cfg(not(debug_assertions))]
+    fn check_magic(&mut self) {}
+
+    #[cfg(debug_assertions)]
+    fn clear_magic(&mut self) {
+        unsafe { std::ptr::write_volatile(&mut self.magic, 0) };
+    }
+    #[cfg(not(debug_assertions))]
+    fn clear_magic(&mut self) {}
```

Comment:
nit: I think rather than two functions it might be more clear as just:

```rust
fn clear_magic(&mut self) {
    #[cfg(debug_assertions)]
    unsafe { std::ptr::write_volatile(&mut self.magic, 0) };
}
```

Reply:
Done